### PR TITLE
 New Put API with support for Append and Overwrite

### PR DESCRIPTION
**Pull request by eolivelli**
New Put API with support for Append and Overwrite
- drop 'append' API
- introduce new PutOptions as specs for Put operations
- implement a new 'concat' API to concatenate two NamedObjects
- implement new listByName API with the only implementation of NamePrefixFilter
- implement a minimal 'ls' command in the CLI to inspect data (only named objects)

### CODE REVIEW

**Comment by eolivelli** [comment]
I think we should not allow to have blobs that belong to more than one NamedObject.
We have to drop the 'append" method in the current form, as it would allow such condition.

My plan is:
- drop "append" in favour of a "contact(name1, name2)", name2 will be merged into name1 and automatically deleted
- introduce PutOptions with a "APPEND" flag, this will allow append in "streaming mode"
- in case of "overwrite" the "old" blobs attached to the NamedObject will be deleted

Please note that "delete" and "concat" areonly a metadata operations, and so they can be done in transaction

File: blobit-core/src/main/java/org/blobit/core/api/BucketHandle.java [hunk 1]
```diff
@@ -39,42 +39,48 @@
      * @param data
      * @return the org.blobit.core.api.PutPromise
      */
-    PutPromise put(String name, byte[] data);
+    default PutPromise put(String name, byte[] data) {
```
Comment (eolivelli) [inline blobit-core/src/main/java/org/blobit/core/api/BucketHandle.java:42]: this should make the API compatible with 0.2 clients (like EmailSuccess)

**Comment by eolivelli** [comment]
@diegosalvi this patch is good to go.

We shuold add the "list" API before 0.3.0, it will make easier to debug problems

**Comment by eolivelli** [comment]
@diegosalvi  I have added to this branch the initial work about 'listing objects'

**Comment by eolivelli** [comment]
Once we merge this change we are ready for cutting the release and I will update the Quickstart guide with the new commands in the CLI.

The structure of the DB schema for the named objects API should be almost stable now.
We don't have support for "directories" but as told offline we need an additional layer in order to emulate a tree, it is not part of the 'named objects' stuff

File: blobit-core/src/main/java/org/blobit/core/api/BucketHandle.java [hunk 2]
```diff
@@ -39,42 +39,48 @@
...
-     * @param name the value of name
+     * @param name the value of name  (optional)
      * @param data array of bytes
      * @param offset offset from which to start writing bytes
      * @param len number of bytes to write
+     * @param putOptions optoins
      * @return the org.blobit.core.api.PutPromise
      */
-    PutPromise put(String name, byte[] data, int offset, int len);
+    PutPromise put(String name, byte[] data, int offset, int len, PutOptions putOptions);
```
Comment (diegosalvi) [inline blobit-core/src/main/java/org/blobit/core/api/BucketHandle.java]: +1 I like methods with array offsets :D

**Review by diegosalvi — APPROVED** [review]

**Comment by diegosalvi** [comment]
CI fails due to spotbugs

**Comment by eolivelli** [comment]
@diegosalvi can you merge this patch ?
remember to clean up the commit message and do not copy all of the secondary commits

**Pull request merged**